pkg/runner/job_runner.go

```diff
@@ -114,77 +146,99 @@ func (jr *JobRunner) Run(j *job.Job) ([][]*job.Report, []*job.Report, error) {
-							j.StateCtx.Warnf("Failed to unlock targets (%v) for job ID %d: %v", targets, j.ID, err)
-						}
-					case <-j.StateCtx.Until(xcontext.ErrPaused):
-						j.StateCtx.Debugf("Received pause request, NOT releasing targets so the job can be resumed")
-						return
-					case <-done:
-						if err := tl.Unlock(j.StateCtx, j.ID, targets); err != nil {
-							j.StateCtx.Warnf("Failed to unlock %d target(s) (%v): %v", len(targets), targets, err)
-						}
-						j.StateCtx.Infof("Unlocked %d target(s) for job ID %d", len(targets), j.ID)
-						return
 					case <-time.After(refreshInterval):
 						// refresh the locks before the timeout expires
-						if err := tl.RefreshLocks(j.StateCtx, j.ID, targets); err != nil {
-							j.StateCtx.Warnf("Failed to refresh %d locks for job ID %d: %v", len(targets), j.ID, err)
+						if err := tl.RefreshLocks(runCtx, j.ID, targets); err != nil {
+							runCtx.Warnf("Failed to refresh %d locks for job ID %d: %v", len(targets), j.ID, err)
 						}
+					// Stop refreshing when told.
+					case <-stopRefresh:
+						close(stoppedRefresh)
+						return
 					}
 				}
 				// refresh locks a bit faster than locking timeout to avoid races
 			}(j, tl, targets, config.LockRefreshTimeout/10*9)
 
 			// Emit events tracking targets acquisition
-			header := testevent.Header{JobID: j.ID, RunID: types.RunID(run + 1), TestName: t.Name}
+			header := testevent.Header{JobID: j.ID, RunID: runID, TestName: t.Name}
 			testEventEmitter := storage.NewTestEventEmitter(header)
 
-			if runErr = jr.emitAcquiredTargets(j.StateCtx, testEventEmitter, targets); runErr == nil {
-				j.StateCtx.Infof("Run #%d: running test #%d for job '%s' (job ID: %d) on %d targets", run+1, idx, j.Name, j.ID, len(targets))
+			if acquired {
+				runErr = jr.emitTargetEvents(runCtx, testEventEmitter, targets, target.EventTargetAcquired)
+			}
+
+			if runErr == nil {
+				runCtx.Infof("Run #%d: running test #%d for job '%s' (job ID: %d) on %d targets",
+					runID, idx, j.Name, j.ID, len(targets))
 				testRunner := NewTestRunner()
-				resumeState, err := testRunner.Run(j.StateCtx, t, targets, j.ID, types.RunID(run+1), nil)
+				var testRunnerState json.RawMessage
+				if resumeState != nil {
+					testRunnerState = resumeState.TestRunnerState
+				}
+				testRunnerState, err := testRunner.Run(ctx, t, targets, j.ID, runID, testRunnerState)
 				if err == xcontext.ErrPaused {
-					j.StateCtx.Debugf("Runner paused, state: %s", string(resumeState))
-					// TODO(rojer): Persist the state.
+					resumeState := &job.PauseEventPayload{
+						Version:         job.CurrentPauseEventPayloadVersion,
+						JobID:           j.ID,
+						RunID:           runID,
+						Targets:         targets,
+						TestRunnerState: testRunnerState,
+					}
+					// Stop refreshing the locks.
+					close(stopRefresh)
+					<-stoppedRefresh
+					// Refresh one last time.
+					if err2 := tl.RefreshLocks(runCtx, j.ID, targets); err2 != nil {
```

Comment:
We probably should lock them for a bit longer when pausing, but that is TODO for the future

Reply:
ack